### PR TITLE
#827 #1679 Improve performance of Request Mapper

### LINKED ISSUE #1679: RequestMapper performance issue: ToByteArray(request.Body)
## Expected Behavior / New Feature

File: [RequestMapper.cs](../blob/develop/src/Ocelot/Request/Mapper/RequestMapper.cs)
Method: [Task&lt;HttpContent&gt; MapContent(HttpRequest request)](/search?q=repo%3AThreeMammals%2FOcelot%20%22Task%3CHttpContent%3E%20MapContent(HttpRequest%20request)%22&type=code)
Line:
```csharp
var content = new ByteArrayContent(await ToByteArray(request.Body));
```
There are performance issues.
Maybe Can change to use `ArrayPool` like YARP, This can reduce GC `byte[]`.

### PR DESCRIPTION
**Pull request by ggnaegi**
## Fixes #1679 #827 
- #1679 
- #827 

## Proposed Changes

  - implementing custom HttpContent, largely inspired by the yarp project.
  https://github.com/microsoft/reverse-proxy/blob/main/src/ReverseProxy/Forwarder/StreamCopyHttpContent.cs

> /// Conclusion: by overriding HttpContent.SerializeToStreamAsync,
    /// we have full control over pumping bytes to the target stream for all protocols
    /// (except Web Sockets, which is handled separately).

@raman-m @RaynaldM I have written a new benchmark, called payload benchmarks (part of PR too). We send various payloads (empty, 25, 50, 100, 1000, 5000, 10000, 20000 json objects), which involves the use of RequestMapper, and evaluate performance and memory usage. I checked that the target client was receiving the payload.

@raman-m @weichaohh As a colleague pointed out, I didn't address the issue of smaller payloads with a length < than the Default Buffer Size. Therefore I have added smaller payloads to the benchmark's payloads. Some are smaller than the Large Object Heap threshold, 85KB. So I modified the copy logic, and made sure that the default buffer size was only used when the promised / announced content length was >= than the Default Buffer Size.

First I have checked the performance of the develop branch code:
image
Figure 1: develop branch code

And then I checked the results of this PR
image
Figure 2:PR code

As you can see, we have a performance gain, but also lower memory usage.
Worst case: 92 MB vs 977 KB for the 20'000 objects payload

### CODE REVIEW

**Comment by ggnaegi** [comment]
@raman-m this is largely inspired by yarp.

**Comment by raman-m** [comment]
Current [build 2178](https://circleci.com/gh/ThreeMammals/Ocelot/2178) has failed!
This is well-known issue #1706

**Review by RaynaldM — APPROVED** [review]

File: src/Ocelot/Request/Mapper/RequestMapper.cs [hunk 1]
```diff
@@ -3,108 +3,92 @@
...
+            if (IsSupportedHeader(header))
+            {
+                requestMessage.Headers.TryAddWithoutValidation(header.Key, header.Value.ToArray());
+            }
+        }
+    }
+
+    private bool IsSupportedHeader(KeyValuePair<string, StringValues> header)
+    {
+        return !_unsupportedHeaders.Contains(header.Key.ToLower());
```
[the remark this answers is not in the document]
Comment (ggnaegi) [reply src/Ocelot/Request/Mapper/RequestMapper.cs:92]: Yes, it makes sense, I will modify that

File: src/Ocelot/Request/Mapper/RequestMapper.cs [hunk 2]
```diff
@@ -3,108 +3,92 @@
...
-                }
-            }
-        }
-    }
+
+namespace Ocelot.Request.Mapper;
+
+public class RequestMapper : IRequestMapper
+{
+    private readonly string[] _unsupportedHeaders = { "host" };
```
[the remark this answers is not in the document]
Comment (ggnaegi) [reply src/Ocelot/Request/Mapper/RequestMapper.cs:11]: Now the question is, do we really need a hashset here, as a const string would be sufficient? My feeling is that this code bit is only here "au cas où"...

[the remark this answers is not in the document]
Comment (ggnaegi) [reply src/Ocelot/Request/Mapper/RequestMapper.cs]: I will keep it like you suggested for the minute.

**Comment by ggnaegi** [comment]
@RaynaldM @raman-m I will provide some benchmarks...

**Comment by raman-m** [comment]
> @RaynaldM @raman-m I will provide some benchmarks...

That will be awesome to attach benchmark results of old version code and new one. Thanks!
Screenshots are better to attach.

**Comment by raman-m** [comment]
I'll review this PR after documentation release... OK?

**Comment by ggnaegi** [comment]
> I'll review this PR after documentation release... OK?

Yeah fine, I will check the performance with 1-2 benchmarks first.

**Comment by weichaohh** [comment]
There is a problem here,Smallest file “100.json” is 139KB, this bigger then 85KB(GC LOP)。soo reduce gc。the arraypool<byte>.shared did noot applicable for greater than 1024*1024.

**Comment by ggnaegi** [comment]
@raman-m Ok, code is now review ready.

**Comment by raman-m** [comment]
Will this PR fix #749  ?

**Review by RaynaldM — APPROVED** [review]

**Review by RaynaldM — APPROVED** [review]

**Comment by raman-m** [comment]
Guys, This PR goes to November release... So, I'll postpone a code review.

**Comment by ggnaegi** [comment]
Yet another interesting library: https://github.com/microsoft/Microsoft.IO.RecyclableMemoryStream

**Comment by raman-m** [comment]
> [Microsoft.IO.RecyclableMemoryStream](https://github.com/microsoft/Microsoft.IO.RecyclableMemoryStream)

Wow! Not bad! 🤩
How did you find this lib? Deep googling? 😁

**Comment by ggnaegi** [comment]
> > [Microsoft.IO.RecyclableMemoryStream](https://github.com/microsoft/Microsoft.IO.RecyclableMemoryStream)
> 
> Wow! Not bad! 🤩 How did you find this lib? Deep googling? 😁

it just appeared out of nowhere, it's magic... Let me think about a possible usage...

**Comment by ggnaegi** [comment]
@raman-m rebase done!

**Comment by raman-m** [comment]
> @raman-m rebase done!

Once again, please! :smiley: 
Conflicts! :sos:

**Comment by raman-m** [comment]
Guys, we're at the finish line... 😃 

@ggnaegi  I see that we have 28 MB payload test in benchmarks.
Some crazy people are using our gateway to download/upload large video files 🤣 

Could we add canonic 10 Mb, 100 Mb and 1 Gb payload tests too please?

File: test/Ocelot.AcceptanceTests/PollyQoSTests.cs [hunk 3]
```diff
@@ -180,11 +180,11 @@ private void GivenThereIsAServiceRunningOn(string url, int statusCode, string re
...
+        public void Dispose()
+        {
+            _serviceHandler?.Dispose();
             _steps.Dispose();
-            GC.SuppressFinalize(this);
-        }
-    }
+            GC.SuppressFinalize(this);
+        }
+    }
```
Comment (raman-m) [inline test/Ocelot.AcceptanceTests/PollyQoSTests.cs:189]: Exclude this file from PR! Zero changes and these tests are not related to the feature changes...

**Review by raman-m — CHANGES_REQUESTED** [review]
Code review 1st round has done.

My concern, it was hard to review because of changed indents in files when namespace definition has changed. So, real changes are not highlighted in some files.
Check your formatting settings once again please in your IDE!

**Comment by ggnaegi** [comment]
@raman-m to be honest I'm not very happy with your review, it's not factual.

File: src/Ocelot/Request/Mapper/RequestMapper.cs [hunk 4]
```diff
@@ -3,108 +3,92 @@
...
-                using (var memStream = new MemoryStream())
-                {
-                    await stream.CopyToAsync(memStream);
-                    return memStream.ToArray();
-                }
-            }
-        }
-    }
+
+namespace Ocelot.Request.Mapper;
```
[the remark this answers is not in the document]
Comment (ggnaegi) [reply src/Ocelot/Request/Mapper/RequestMapper.cs:7]: I asked you about it. I prefer the new notation, less brackets

File: src/Ocelot/Request/Mapper/StreamHttpContent.cs [hunk 5]
```diff
@@ -0,0 +1,118 @@
+﻿using Microsoft.AspNetCore.Http;
+using System.Buffers;
+
+namespace Ocelot.Request.Mapper;
+
+public class StreamHttpContent : HttpContent
+{
+    private const int DefaultBufferSize = 65536;
```
[the remark this answers is not in the document]
Comment (ggnaegi) [reply src/Ocelot/Request/Mapper/StreamHttpContent.cs:8]: Yes

File: src/Ocelot/Request/Mapper/StreamHttpContent.cs [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+    {
+        await CopyAsync(_context.Request.Body, stream, Headers.ContentLength ?? UnknownLength, false,
+            CancellationToken.None);
+    }
+
+    protected override bool TryComputeLength(out long length)
+    {
+        length = -1;
+        return false;
+    }
```
[the remark this answers is not in the document]
Comment (ggnaegi) [reply src/Ocelot/Request/Mapper/StreamHttpContent.cs:34]: Yes, but why would we need that in the API-Gateway, it must be quick, why would you try to compute the length?

File: src/Ocelot/Request/Mapper/StreamHttpContent.cs [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+
+    // this is used internally by HttpContent.ReadAsStreamAsync(...)
+    protected override Task<Stream> CreateContentReadStreamAsync()
+    {
+        // Nobody should be calling this...
+        throw new NotImplementedException();
+    }
```
[the remark this answers is not in the document]
Comment (ggnaegi) [reply src/Ocelot/Request/Mapper/StreamHttpContent.cs:41]: Same as before.

File: src/Ocelot/Request/Mapper/StreamHttpContent.cs [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+        // for smaller payloads, avoid allocating a buffer that is larger than the announced content length
+        var minBufferSize = announcedContentLength != UnknownLength && announcedContentLength < DefaultBufferSize
+            ? (int)announcedContentLength
+            : DefaultBufferSize;
+
+        var buffer = ArrayPool<byte>.Shared.Rent(minBufferSize);
+        long contentLength = 0;
+        try
+        {
+            while (true)
```
[the remark this answers is not in the document]
Comment (ggnaegi) [reply src/Ocelot/Request/Mapper/StreamHttpContent.cs:55]: it's not an endless loop, it breaks when read == 0

File: test/Ocelot.AcceptanceTests/ContentTests.cs [hunk 9]
```diff
@@ -91,7 +91,7 @@ public void should_add_content_type_and_content_length_headers()
                 .And(x => _steps.GivenThePostHasContentType(contentType))
                 .When(x => _steps.WhenIPostUrlOnTheApiGateway("/"))
                 .Then(x => _steps.ThenTheStatusCodeShouldBe(HttpStatusCode.Created))
-                .And(x => ThenTheContentLengthIs(11))
+                // .And(x => ThenTheContentLengthIs(11))
```
[the remark this answers is not in the document]
Comment (ggnaegi) [reply test/Ocelot.AcceptanceTests/ContentTests.cs:94]: Yes should be removed, plain and simple

File: test/Ocelot.AcceptanceTests/PollyQoSTests.cs (hunk 3 above)
Comment (ggnaegi) [reply test/Ocelot.AcceptanceTests/PollyQoSTests.cs:189]: Maybe merge is responsible?

File: test/Ocelot.Benchmarks/Payloads/100.json [hunk 10]
```diff
@@ -0,0 +1,4502 @@
...
+    "guid": "5bdc2073-f4d9-4fce-9526-f715cd2f63d0",
+    "isActive": true,
+    "balance": "$2,092.91",
+    "picture": "http://placehold.it/32x32",
+    "age": 32,
+    "eyeColor": "green",
+    "name": "Valencia Day",
+    "gender": "male",
+    "company": "KAGE",
+    "email": "valenciaday@kage.com",
```
[the remark this answers is not in the document]
Comment (ggnaegi) [reply test/Ocelot.Benchmarks/Payloads/100.json:14]: it's all fake

File: test/Ocelot.UnitTests/Request/Mapper/RequestMapperTests.cs [hunk 11]
```diff
@@ -217,7 +216,6 @@ public void should_not_add_content_headers()
                 .When(_ => WhenMapped())
                 .Then(_ => ThenNoErrorIsReturned())
                 .And(_ => ThenTheMappedRequestHasContentTypeHeader("application/json"))
-                .And(_ => ThenTheMappedRequestHasContentSize("This is my content".Length))
```
[the remark this answers is not in the document]
Comment (ggnaegi) [reply test/Ocelot.UnitTests/Request/Mapper/RequestMapperTests.cs:220]: If you want, we can compute it, I'm not sure about it though

File: src/Ocelot/Request/Mapper/RequestMapper.cs [hunk 12]
```diff
@@ -3,108 +3,92 @@
...
+        if (!string.IsNullOrEmpty(request.ContentType))
+        {
+            content.Headers
+                .TryAddWithoutValidation("Content-Type", new[] { request.ContentType });
+        }
+
+        AddHeaderIfExistsOnRequest("Content-Language", content, request);
+        AddHeaderIfExistsOnRequest("Content-Location", content, request);
+        AddHeaderIfExistsOnRequest("Content-Range", content, request);
+        AddHeaderIfExistsOnRequest("Content-MD5", content, request);
```
[the remark this answers is not in the document]
Comment (ggnaegi) [reply src/Ocelot/Request/Mapper/RequestMapper.cs:53]: We can remove this too

File: test/Ocelot.Benchmarks/Payloads/100.json (hunk 10 above)
[the remark this answers is not in the document]
Comment (raman-m) [reply test/Ocelot.Benchmarks/Payloads/100.json:14]: So, all emails and phones are fake. OK
Not an issue!

File: src/Ocelot/Request/Mapper/StreamHttpContent.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (raman-m) [reply src/Ocelot/Request/Mapper/StreamHttpContent.cs:55]: Lines [L90-L93](https://github.com/ggnaegi/Ocelot/blob/features/1679-request-mapper-performance-issues/src/Ocelot/Request/Mapper/StreamHttpContent.cs#L90-L93)
Is this the ending loop?
```csharp
if (announcedContentLength == UnknownLength || contentLength == announcedContentLength)
{
    return;
}
```

File: src/Ocelot/Request/Mapper/StreamHttpContent.cs [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+
+    private static async Task CopyAsync(Stream input, Stream output, long announcedContentLength,
+        bool autoFlush, CancellationToken cancellation)
+    {
+        // for smaller payloads, avoid allocating a buffer that is larger than the announced content length
+        var minBufferSize = announcedContentLength != UnknownLength && announcedContentLength < DefaultBufferSize
+            ? (int)announcedContentLength
+            : DefaultBufferSize;
+
+        var buffer = ArrayPool<byte>.Shared.Rent(minBufferSize);
```
[the remark this answers is not in the document]
Comment (raman-m) [reply src/Ocelot/Request/Mapper/StreamHttpContent.cs:51]: Very significant conclusion! Because the logic says "Larger buffer should show better performance".
But in reality it doesn't work for large streams... and streaming scenarios.
🆗 In my opinion, this is a challenge to guess the best buffer size. We can only define very common buffer size...
Some routes can proxy small bodies, some routes can pump large content in streaming scenario...
I would say, it is a bit wrong to define one common buffer size for all routes of Ocelot app instance...
Did you get my thought?

File: src/Ocelot/Request/Mapper/StreamHttpContent.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (raman-m) [reply src/Ocelot/Request/Mapper/StreamHttpContent.cs:34]: Aha! Finally got it! 😄  We switch off content length computing... Is that the same than our lovely Bla-Bla gateway does?

Hey @RaynaldM , what do you think?

File: src/Ocelot/Request/Mapper/RequestMapper.cs (hunk 12 above)
[the remark this answers is not in the document]
Comment (raman-m) [reply src/Ocelot/Request/Mapper/RequestMapper.cs:53]: Removing headers means that downstream service can respond with error or have abnormal behavior.
Truly speaking we have to forward old headers, a few of them should be hidden...

I don't agree to remove now.
Not an issue...

File: src/Ocelot/Request/Mapper/RequestMapper.cs [hunk 14]
```diff
@@ -3,108 +3,92 @@
...
-        }
-    }
+
+namespace Ocelot.Request.Mapper;
+
+public class RequestMapper : IRequestMapper
+{
+    private readonly HashSet<string> _unsupportedHeaders = new(StringComparer.OrdinalIgnoreCase) { "host" };
+
+    public Task<Response<HttpRequestMessage>> Map(HttpRequest request, DownstreamRoute downstreamRoute)
```
[the remark this answers is not in the document]
Comment (ggnaegi) [reply src/Ocelot/Request/Mapper/RequestMapper.cs:13]: @raman-m we can try, but we are moving the problem to ```DownstreamRequestInitialiserMiddleware``` At least we can remove the Task

File: src/Ocelot/Request/Mapper/RequestMapper.cs [hunk 15]
```diff
@@ -3,108 +3,92 @@
...
+            {
+                Content = MapContent(request),
+                Method = MapMethod(request, downstreamRoute),
+                RequestUri = MapUri(request),
+                Version = downstreamRoute.DownstreamHttpVersion,
+            };
+
+            MapHeaders(request, requestMessage);
+
+            return Task.FromResult<Response<HttpRequestMessage>>(new OkResponse<HttpRequestMessage>(requestMessage));
```
[the remark this answers is not in the document]
Comment (ggnaegi) [reply src/Ocelot/Request/Mapper/RequestMapper.cs:27]: Yes, so I recommend changing the signature returning Response<HttpRequestMessage> instead of Task<...>

File: src/Ocelot/Request/Mapper/RequestMapper.cs (hunk 14 above)
[the remark this answers is not in the document]
Comment (ggnaegi) [reply src/Ocelot/Request/Mapper/RequestMapper.cs:13]: @raman-m Ok, agree at the end, but we will need to review the error handling later...

File: src/Ocelot/Request/Mapper/RequestMapper.cs [hunk 16]
```diff
@@ -3,108 +3,92 @@
...
+        }
+        catch (Exception ex)
+        {
+            return Task.FromResult<Response<HttpRequestMessage>>(new ErrorResponse<HttpRequestMessage>(new UnmappableRequestError(ex)));
+        }
+    }
+
+    private static HttpContent MapContent(HttpRequest request)
+    {
+        if (request.Body is null or { CanSeek: true, Length: <= 0 })
```
[the remark this answers is not in the document]
Comment (ggnaegi) [reply src/Ocelot/Request/Mapper/RequestMapper.cs:33]: @raman-m it means if CanSeek = true and Length <= 0, there can't be any content.. If CanSeek is false, then you can't possibly know the length, and even if Length is 0

**Comment by ggnaegi** [comment]
@raman-m as you wished, it's ready for a new review...

**Comment by ggnaegi** [comment]
@raman-m I can't push the 1GB payload, it's 100MB maximum

File: src/Ocelot/Request/Mapper/RequestMapper.cs [hunk 17]
```diff
@@ -3,108 +3,92 @@
...
+        {
+            return null;
+        }
+
+        var content = new StreamHttpContent(request.HttpContext);
+
+        if (!string.IsNullOrEmpty(request.ContentType))
+        {
+            content.Headers
+                .TryAddWithoutValidation("Content-Type", new[] { request.ContentType });
```
[the remark this answers is not in the document]
Comment (ggnaegi) [reply src/Ocelot/Request/Mapper/RequestMapper.cs:50]: ok

File: src/Ocelot/Request/Mapper/StreamHttpContent.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (ggnaegi) [reply src/Ocelot/Request/Mapper/StreamHttpContent.cs:37]: you know who

File: src/Ocelot/Request/Mapper/StreamHttpContent.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (ggnaegi) [reply src/Ocelot/Request/Mapper/StreamHttpContent.cs:47]: Yes, it's the rule of thumb here...

File: src/Ocelot/Request/Mapper/StreamHttpContent.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (ggnaegi) [reply src/Ocelot/Request/Mapper/StreamHttpContent.cs:51]: yes

File: src/Ocelot/Request/Mapper/RequestMapper.cs [hunk 18]
```diff
@@ -3,108 +3,92 @@
...
+
+    private static HttpMethod MapMethod(HttpRequest request, DownstreamRoute downstreamRoute)
+    {
+        return !string.IsNullOrEmpty(downstreamRoute?.DownstreamHttpMethod) ? new HttpMethod(downstreamRoute.DownstreamHttpMethod) : new HttpMethod(request.Method);
+    }
+
+    private static Uri MapUri(HttpRequest request)
+    {
+        return new Uri(request.GetEncodedUrl());
+    }
```
[the remark this answers is not in the document]
Comment (ggnaegi) [reply src/Ocelot/Request/Mapper/RequestMapper.cs]: Yes, done.

File: src/Ocelot/Request/Mapper/RequestMapper.cs (hunk 15 above)
[the remark this answers is not in the document]
Comment (ggnaegi) [reply src/Ocelot/Request/Mapper/RequestMapper.cs]: I choose to go a step further here and return an HttpRequestMessage object instead

**Comment by raman-m** [comment]
Ah, sorry... To keep this 100mb payload as a file in git?
If Yes that will be not good.

File: src/Ocelot/Request/Mapper/RequestMapper.cs (hunk 14 above)
[the remark this answers is not in the document]
Comment (ggnaegi) [reply src/Ocelot/Request/Mapper/RequestMapper.cs]: and even better ```public HttpRequestMessage Map(HttpRequest request, DownstreamRoute downstreamRoute)```

File: src/Ocelot/Request/Mapper/StreamHttpContent.cs [hunk 19]
```diff
@@ -0,0 +1,118 @@
...
+                    buffer = ArrayPool<byte>.Shared.Rent(minBufferSize);
+                }
+
+                var read = await input.ReadAsync(buffer.AsMemory(), cancellation);
+                contentLength += read;
+
+                // Normally this is enforced by the server, but it could get out of sync if something in the proxy modified the body.
+                if (announcedContentLength != UnknownLength && contentLength > announcedContentLength)
+                {
+                    throw new InvalidOperationException("More bytes received than the specified Content-Length.");
```
[the remark this answers is not in the document]
Comment (ggnaegi) [reply src/Ocelot/Request/Mapper/StreamHttpContent.cs:80]: Ok, done

File: src/Ocelot/Request/Mapper/StreamHttpContent.cs [hunk 20]
```diff
@@ -0,0 +1,118 @@
...
+                }
+
+                // end of the source stream.
+                if (read == 0)
+                {
+                    if (announcedContentLength == UnknownLength || contentLength == announcedContentLength)
+                    {
+                        return;
+                    }
```
[the remark this answers is not in the document]
Comment (ggnaegi) [reply src/Ocelot/Request/Mapper/StreamHttpContent.cs:89]: I don't think so, because of this:
```c#
var read = await input.ReadAsync(buffer.AsMemory(), cancellation);
contentLength += read;
```

File: src/Ocelot/Request/Mapper/StreamHttpContent.cs [hunk 21]
```diff
@@ -0,0 +1,118 @@
...
+                    else
+                    {
+                        throw new InvalidOperationException(
+                            $"Sent {contentLength} request content bytes, but Content-Length promised {announcedContentLength}.");
```
[the remark this answers is not in the document]
Comment (ggnaegi) [reply src/Ocelot/Request/Mapper/StreamHttpContent.cs:93]: Nice to add, but it's imho not part of current scope. So I would skip this for the minute, since we are not using Exception.Data in the solution yet.

**Comment by ggnaegi** [comment]
> Ah, sorry... To keep this 100mb payload as a file in git? If Yes that will be not good.

You asked for bigger payloads. 100mb is fine, but 1gb is too big for github. So it's not part of the benchmarks now. I hope it's fine for you... Or I could create the payload on the fly, it's maybe better for the end users.

**Comment by raman-m** [comment]
Yes, We should refactor benchmarks to operate and process in-memory objects rather than file system ones.
But this is not critical now...
Is it difficult to refactor to switch from file to in-memory or stream objects?
If Yes, please ignore this problem now.
Adding 100mb file to git is bad for sure...

It seems it is time to create Assets branch in repo to save static files aka blobs...
Will do that soon ...

**Comment by ggnaegi** [comment]
@raman-m So for the 10 100 and 1024Mb payloads, i have provided a method to generate dummy content

**Comment by raman-m** [comment]
I've got another idea...
It is totally fine to generate large files during test setup step rather than keep them as static files in Git...
But seems you're going to refactor in favor of memory streams...

**Comment by ggnaegi** [comment]
@raman-m no no, I'm creating the payloads when starting the benchmarks for the first time. No memory stream. It's pushed, you can review! 
```c#
/// <summary>
    /// Generates a dummy payload of the given size in MB.
    /// Avoiding to maintain a large file in the repository.
    /// </summary>
    /// <param name="size">The file size in MB.</param>
    /// <param name="payloadPath">The path to the payload file.</param>
    /// <returns>The payload as byte array.</returns>
    /// <exception cref="ArgumentNullException">Throwing an exception if the payload path is null.</exception>
    private byte[] GenerateDummyPayload(int size, string payloadPath)
    {
        if (payloadPath == null)
        {
            throw new ArgumentNullException(nameof(payloadPath));
        }

        if (File.Exists(payloadPath))
        {
            return File.ReadAllBytes(payloadPath);
        }

        var newFile = new FileStream(payloadPath, FileMode.CreateNew);
        newFile.Seek(size * 1024L * 1024, SeekOrigin.Begin);
        newFile.WriteByte(0);
        newFile.Close();

        return File.ReadAllBytes(payloadPath);
    }
```

File: test/Ocelot.Benchmarks/PayloadBenchmarks.cs [hunk 22]
```diff
@@ -0,0 +1,236 @@
...
+        {
+            File.ReadAllBytes(Path.Combine(baseDirectory, "Payloads", "10000.json")),
+            true,
+        };
+
+        yield return new object[]
+        {
+            File.ReadAllBytes(Path.Combine(baseDirectory, "Payloads", "20000.json")),
+            true,
+        };
```
[the remark this answers is not in the document]
Comment (ggnaegi) [reply test/Ocelot.Benchmarks/PayloadBenchmarks.cs:122]: It's easy to create a dummy dat file... I could try with Bogus https://github.com/bchavez/Bogus to generate the json files

File: test/Ocelot.Benchmarks/Ocelot.Benchmarks.csproj [hunk 23]
```diff
@@ -38,6 +38,33 @@
...
+    </None>
+    <None Update="Payloads\1000.json">
+      <CopyToOutputDirectory>Always</CopyToOutputDirectory>
+    </None>
+    <None Update="Payloads\10000.json">
+      <CopyToOutputDirectory>Always</CopyToOutputDirectory>
+    </None>
+    <None Update="Payloads\20000.json">
+      <CopyToOutputDirectory>Always</CopyToOutputDirectory>
+    </None>
```
[the remark this answers is not in the document]
Comment (ggnaegi) [reply test/Ocelot.Benchmarks/Ocelot.Benchmarks.csproj:53]: Ok, I will try with bogus, you're right

File: test/Ocelot.Benchmarks/PayloadBenchmarks.cs [hunk 24]
```diff
@@ -0,0 +1,236 @@
...
+        {
+            throw new ArgumentNullException(nameof(payloadPath));
+        }
+
+        if (File.Exists(payloadPath))
+        {
+            return File.ReadAllBytes(payloadPath);
+        }
+
+        var newFile = new FileStream(payloadPath, FileMode.CreateNew);
```
[the remark this answers is not in the document]
Comment (ggnaegi) [reply test/Ocelot.Benchmarks/PayloadBenchmarks.cs:163]: yes, it's better.

File: test/Ocelot.Benchmarks/Ocelot.Benchmarks.csproj (hunk 23 above)
[the remark this answers is not in the document]
Comment (raman-m) [reply test/Ocelot.Benchmarks/Ocelot.Benchmarks.csproj:53]: I'm not sure about external libs...
Maybe it is OK for testing projects to reference to external libs, but it is forbidden for Ocelot main project.

File: test/Ocelot.Benchmarks/PayloadBenchmarks.cs (hunk 22 above)
[the remark this answers is not in the document]
Comment (raman-m) [reply test/Ocelot.Benchmarks/PayloadBenchmarks.cs:122]: Not bad lib! Have you tried it in another projects?

What about another test data generating approach...
We can take 1 MB file from Git and copy it into a file stream as many times as a test requires. For example for generating 10 MB temp data file with test data we could copy 1 MB file exactly 10 times.
Sounds good?
Canonic 1 MB file should contain good test data for sure...

File: test/Ocelot.Benchmarks/PayloadBenchmarks.cs (hunk 24 above)
[the remark this answers is not in the document]
Comment (raman-m) [reply test/Ocelot.Benchmarks/PayloadBenchmarks.cs:163]: Waiting for the fix... 😉

File: test/Ocelot.Benchmarks/PayloadBenchmarks.cs [hunk 25]
```diff
@@ -0,0 +1,236 @@
...
+            .ConfigureAppConfiguration((hostingContext, config) =>
+            {
+                config
+                    .SetBasePath(hostingContext.HostingEnvironment.ContentRootPath)
+                    .AddJsonFile("appsettings.json", true, true)
+                    .AddJsonFile($"appsettings.{hostingContext.HostingEnvironment.EnvironmentName}.json", true, true)
+                    .AddJsonFile("ocelot.json", false, false)
+                    .AddEnvironmentVariables();
+            })
+            .ConfigureKestrel((_, hostingOptions) => { hostingOptions.Limits.MaxRequestBodySize = 1100000000; })
```
[the remark this answers is not in the document]
Comment (raman-m) [reply test/Ocelot.Benchmarks/PayloadBenchmarks.cs:186]: Great! Going to add one extra commit with [gotchas.rst](https://github.com/ggnaegi/Ocelot/blob/features/1679-request-mapper-performance-issues/docs/introduction/gotchas.rst) update...

File: test/Ocelot.Benchmarks/PayloadBenchmarks.cs (hunk 22 above)
[the remark this answers is not in the document]
Comment (raman-m) [reply test/Ocelot.Benchmarks/PayloadBenchmarks.cs:122]: But my main concern was/is that we added some large static files into Git-database. That's bad!
Git history as Git-database will keep these files forever having total Git DB size increased by static files sizes...
It seems it is too late... we cannot do decrease. Or Am I wrong? When we will merge to develop via squash option these files will not be added to `develop` & `main` branches, right?

File: test/Ocelot.Benchmarks/PayloadBenchmarks.cs (hunk 25 above)
[the remark this answers is not in the document]
Comment (raman-m) [reply test/Ocelot.Benchmarks/PayloadBenchmarks.cs:186]: Added [docs](https://github.com/ThreeMammals/Ocelot/pull/1724/files/befaab84e8ad2c3b407875c8555a920661bc569d..35d115219908793144ce38e8200d6c07886104d1)!
Fixed in https://github.com/ThreeMammals/Ocelot/pull/1724/commits/35d115219908793144ce38e8200d6c07886104d1 ✔️

**Comment by raman-m** [comment]
Gui, [gotchas.rst](https://github.com/ggnaegi/Ocelot/blob/features/1679-request-mapper-performance-issues/docs/introduction/gotchas.rst) looks good now. 😃
It is a part of docs updating.
This PR changes internal logic of content forwarder, so we should not have many updates in docs. I will think more what we can update in docs... but seems not much

File: src/Ocelot/Request/Mapper/StreamHttpContent.cs [hunk 26]
```diff
@@ -0,0 +1,113 @@
...
+    protected override bool TryComputeLength(out long length)
+    {
+        length = -1;
+        return false;
+    }
+
+    // This is used internally by HttpContent.ReadAsStreamAsync(...)
+    protected override Task<Stream> CreateContentReadStreamAsync()
+    {
+        // Nobody should be calling this...
```
Comment (RaynaldM) [inline src/Ocelot/Request/Mapper/StreamHttpContent.cs:35]: @ggnaegi 
unfortunately, it can be called from a delegateHandler, like this for example
```
 // get upstream content
 if (request.Content != null)
 {
     await using var upstreamContentStream = await request.Content.ReadAsStreamAsync(cancellationToken);
     using var reader = new StreamReader(upstreamContentStream);
      ...
 }
```
Can't you leave the original?

**Review by RaynaldM — APPROVED** [review]

**Review by RaynaldM — APPROVED** [review]

**Review by raman-m — APPROVED** [review]
Ready for delivery! ✔️ 

Thanks Gui for your efforts! 😍 
This will be the brilliant feature of November'23 release! 💎

**Comment by RaynaldM** [comment]
We've had this PR in production for 5 days (plus a trial run a few days earlier), it's working well, and we haven't had any Out of Memory issues, or even a slightly lower memory footprint from the app.

**Comment by raman-m** [comment]
> @RaynaldM commented [on Dec 14](https://github.com/ThreeMammals/Ocelot/pull/1724#issuecomment-1855550188)

🆗 Good to know! 👍 
Please, watch the system in production further...  OOM and Docker container vs K8s pods restarts also?!
